baseplate/clients/redis_cluster.py

```diff
@@ -0,0 +1,225 @@
+import random
+
+from typing import Any
+from typing import Dict
+
+import rediscluster  # type: ignore
+
+from rediscluster.pipeline import ClusterPipeline  # type: ignore
+
+from baseplate import Span
+from baseplate.clients import ContextFactory
+from baseplate.lib import config
+from baseplate.lib import metrics
+
+
+# We want to be able to combine blocking behaviour with the ability to read from replicas
+# Unfortunately this is not provide as-is so we cmobine two connection pool classes to provide
+# the desired behaviour.
+class ClusterWithReadReplicasBlockingConnectionPool(rediscluster.ClusterBlockingConnectionPool):
+    # pylint: disable=arguments-differ
+    def get_node_by_slot(self, slot: int, read_command: bool = False) -> Dict[str, Any]:
+        """
+        Get a random node from the slot, including master
+        """
+        nodes_in_slot = self.nodes.slots[slot]
+        if read_command:
+            random_index = random.randrange(1, len(nodes_in_slot))
+            return nodes_in_slot[random_index]
+
+        return nodes_in_slot[0]
```

Comment:
Am I understanding right that the first node is always the primary and the rest are replicas? If so, would something like this make that a little more self-explanatory?

```suggestion
        primary, *replicas = self.nodes.slots[slot]
        if read_command:
            return random.choice(replicas)
        return primary
```

Edge case check: is there ever a situation where we'd have zero replicas?

Reply:
actually it's definitely possible, yes, let me rework it so it can handle all cases